### ISSUE TITLE
Setting custom parameters to scenario to access them in After hook

### ISSUE THREAD

**Issue by mannejkumar** [cucumber/cucumber-jvm#987]
I am trying to integrate cucumber with testrail. To send the result to testrail i need the test case id. I am getting scenario.name and scenario.id in After hook. But i need to pass custom id from feature file and set it to scenario object. for example i need to pass TC01 while running first scenario. then i get this id in Afterhook to send the http request to test rail along with id. Is there any possibility to set the scenario id manually for each scenario

**Comment by tcwmj** [comment]
You might not to set the custom parameter to the hook methods.There is another approach to setup a map with the scenario as a key and your parameter objects as value, you can get your parameters by the key in any methods.

**Comment by mannejkumar** [comment]
where can i find this code? actually after executing scenario, before taking screenshot in "after" method  i want to get the test case id to send the result to test rail `After { scenario ->
       // send test result to testrail.  by using below code
      Map data = new HashMap();
data.put("status_id", new Integer(1));
data.put("comment", "This test worked fine!");
//need to get test case id from feature file .
JSONObject r = (JSONObject) client.sendPost("add_result_for_case/1/1", data);

	// embed screenshot into cucumber report
	if(scenario.failed) {
		try {
			scenario.embed(theBrowser.driver.getScreenshotAs(OutputType.BYTES), "image/png")
		} catch(WebDriverException e) {
			// sometime firefox runs out of memory trying to take a screenshot, not a big deal so ignore
		} catch(MissingMethodException e) {
			// HTMLUnit doesn't support screenshots
		}
	}

}`

**Comment by aslakhellesoy** [comment]
Scenarios have a name, but they don't have an id. Cucumber does not require Scenario names to be unique, so even if you did have access to the scenario name (and use that as the id), there would be a risk that there would be several scenarios with the same name.

Are you trying to link results of Cucumber scenarios with TestRail test cases? What if you could put the TestRail test case id as a tag above the scenario - would that work for you if the scenario allowed you to get the tags of a scenario?

**Comment by yghumade** [comment]
Hello, 
Is this implemented?, even i have the same requirement where i want to pass or tag the test case id(e.g tc1234) to feature file which relates to test case id in testrail  and then fetch that in definition file or other utility class such as test result upload class which will be uploading the results back to testrail with its result(pass/fail)

**Comment by mpkorstanje** [comment]
Reporting and related concerns are not part of the test. Rather then using the Before and After hooks it would be better to use a Formatter.

Either:
1. Use the JSONFormatter to output json for each test. After the tests are done gather up the json files and upload them to TestRail. This can be step in your build process like a maven plugin, post build script.

2. But I'd actually recommend writing your own reporter for this. The reporter has access to just about everything. Not just the tags, and step names and test results but also individual steps, their full text, any comments. Once 2.0.0 is done this will be especially easy.
